File: cpp_1-9/4.cc
/*
 * C++
 * leetcode 4
 * url: https://leetcode-cn.com/problems/median-of-two-sorted-arrays/
 * 奇数个只需要一个，偶数个需要两个
 * 不管如何，都用两个int 来装答案，如果是奇数个，那第二个 = 第一个
 * 用i，j当指针指向两个vector下一个待比较的index
 * get表示已经取的的小数字的个数
 * set t1部分用来设定第一个答案
 * set get部分用来设定第二个答案
 * 最后返回两数之和除以2.0
 */

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
    	int size1 =nums1.size(), size2 = nums2.size();
    	int t1 = (size1+size2)/2, i =0, j =0, get = 0;
    	bool needSecond = false;
    	if((size1+size2)%2==0){
    		needSecond = true;
    		t1--;
    	}

    	while(get<t1){
    		if(i ==size1)
    			j++;
    		else if(j == size2)
    			i++;
    		else{
    			if(nums1[i] <nums2[j]){
    				i++;
    			}else{
    				j++;
    			}
    		}
    		get++;
    	}
    	// set t1
		if(i ==size1){
			t1 = nums2[j];
			j++;
		}else if(j == size2){
			t1 = nums1[i];
			i++;
		}else{
			if(nums1[i] <nums2[j]){
				t1 = nums1[i];
				i++;
			}else{
				t1 = nums2[j];
				j++;
			}
		}
		// set get
		if(!needSecond){
			get = t1;
		}else{
			if(i ==size1){
				get = nums2[j];
			}else if(j == size2){
				get = nums1[i];
			}else{
				if(nums1[i] <nums2[j]){
					get = nums1[i];
				}else{
					get = nums2[j];
				}
			}
		}

		return (t1+get)/2.0;
    }
};

File: Go/Go_1-9/4.go
// Go
// leetcode 4
// https://leetcode.cn/problems/median-of-two-sorted-arrays/
// 左右指针，略蛋疼

func findMedianSortedArrays(nums1 []int, nums2 []int) (res float64) {
	n1, n2 := len(nums1), len(nums2)
	pass := (len(nums1) + len(nums2) - 1) / 2
	l_pass, r_pass := 0, 0
	l1, l2 := 0, 0
	r1, r2 := n1 - 1, n2 - 1
	for l_pass < pass {
		if l1 > r1 {
			// l1 炸了
			l2 += pass - l_pass
			l_pass = pass
			continue
		}

		if l2 > r2 {
			l1 += pass - l_pass
			l_pass = pass
			continue
		}

		if nums1[l1] > nums2[l2] {
			l2++
		} else {
			l1++
		}
		l_pass++
	}

	for r_pass < pass {
		if l1 > r1 {
			r2 -= pass - r_pass
			r_pass = pass
			continue
		}

		if l2 > r2 {
			r1 -= pass - r_pass
			r_pass = pass
			continue
		}

		if nums1[r1] > nums2[r2] {
			r1--
		} else {
			r2--
		}
		r_pass++
	}
	// fmt.Println(l1,r1,l2,r2)
	if l1 < r1 {
		return (float64(nums1[l1]) + float64(nums1[r1])) / 2
	}

	if l2 < r2 {
		return (float64(nums2[l2]) + float64(nums2[r2])) / 2
	}

	if l1 == r1 {
		res = float64(nums1[l1])
	} else {
		res = float64(nums2[l2])
	}

	if l2 == r2 {
		res = (res + float64(nums2[l2])) / 2
	}

	return res
}